orttraining/orttraining/core/framework/triton/triton_op_executor.cc

```diff
@@ -0,0 +1,132 @@
+              executor_by_name != nullptr && executor_by_onnx_.get() == nullptr && executor_by_onnx != nullptr);
+  Py_INCREF(config_getter);
+  Py_INCREF(executor_by_name);
+  Py_INCREF(executor_by_onnx);
+  PythonObjectPtr config_getter_ptr(config_getter, PythonObjectDeleter);
+  config_getter_ = std::move(config_getter_ptr);
+  PythonObjectPtr executor_by_name_ptr(executor_by_name, PythonObjectDeleter);
+  executor_by_name_ = std::move(executor_by_name_ptr);
+  PythonObjectPtr executor_by_onnx_ptr(executor_by_onnx, PythonObjectDeleter);
+  executor_by_onnx_ = std::move(executor_by_onnx_ptr);
+}
+
+std::string TritonOpExecutor::GetConfigJson() {
+  ORT_ENFORCE(IsInitialized());
+  // Python-related calls should happen only if guard is alive.
+  GilGuard guard;
+  PythonObjectPtr ret(PyObject_CallObject(config_getter_.get(), nullptr), PythonObjectDeleter);
+  char* buffer = nullptr;
+  Py_ssize_t length;
+  buffer = const_cast<char*>(PyUnicode_AsUTF8AndSize(ret.get(), &length));
+  return std::string(buffer, length);
+}
+
+void TritonOpExecutor::ExecuteByOnnx(int64_t onnx_key, const std::string& onnx_string,
+                                     const InlinedVector<const OrtValue*>& inputs, InlinedVector<OrtValue>& outputs,
+                                     const InlinedHashSet<size_t>& bool_outputs) {
+  ORT_ENFORCE(IsInitialized());
+  // Python-related calls should happen only if guard is alive.
+  GilGuard guard;
+  PythonObjectPtr args(PyTuple_New(static_cast<Py_ssize_t>(2 + inputs.size())), PythonObjectDeleter);
+  ORT_ENFORCE(args, "Failed to create args.");
+  PyTuple_SetItem(args.get(), 0, PyLong_FromLongLong(static_cast<long long>(onnx_key)));
+  PyTuple_SetItem(args.get(), 1, PyBytes_FromStringAndSize(onnx_string.c_str(), onnx_string.size()));
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    if (!inputs[i]) {
+      PyTuple_SetItem(args.get(), static_cast<Py_ssize_t>(2 + i), Py_None);
+      Py_INCREF(Py_None);
+    } else {
+      PyTuple_SetItem(args.get(), static_cast<Py_ssize_t>(2 + i), torch::ToDlpack(*inputs[i]));
+    }
+  }
+
+  PythonObjectPtr ret(PyObject_CallObject(executor_by_onnx_.get(), args.get()), PythonObjectDeleter);
+  if (ret == nullptr) {
+    PyErr_Print();
+    ORT_THROW("Python function execution fails with the above information.");
+  }
+  ORT_ENFORCE(ret.get() != Py_None);
+  if (PyTuple_Check(ret.get())) {
+    for (size_t i = 0; i < static_cast<size_t>(PyTuple_Size(ret.get())); ++i) {
+      outputs.emplace_back(torch::FromDlpack(PyTuple_GetItem(ret.get(), static_cast<Py_ssize_t>(i)),
+                                             bool_outputs.find(i) != bool_outputs.end()));
+    }
+  } else {
+    outputs.emplace_back(torch::FromDlpack(ret.get(), bool_outputs.find(0) != bool_outputs.end()));
+  }
+}
+
+void TritonOpExecutor::ExecuteByFuncName(const std::string& func_name, const InlinedVector<const OrtValue*>& inputs,
```

Comment:
Can you explain when will this be used?

Reply:
Added some simple comment into the header file. ExecuteByOnnx is the codegen logic for onnx graph. For manual kernel execution and Op tuning implementation, it's called from this ExecuteByName, the func_name is manual written Python function name.